cli/aws_provider_patch.go

```diff
@@ -0,0 +1,179 @@
+package cli
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/gruntwork-io/terragrunt/errors"
+	"github.com/gruntwork-io/terragrunt/options"
+	"github.com/gruntwork-io/terragrunt/util"
+	"github.com/hashicorp/hcl2/hcl"
+	"github.com/hashicorp/hcl2/hcl/hclsyntax"
+	"github.com/hashicorp/hcl2/hclwrite"
+	"github.com/mattn/go-zglob"
+	"github.com/zclconf/go-cty/cty"
+	"io/ioutil"
+	"path/filepath"
+)
+
+func applyAwsProviderPatch(terragruntOptions *options.TerragruntOptions) error {
+	if len(terragruntOptions.AwsProviderPatchOverrides) == 0 {
+		return errors.WithStackTrace(MissingOverrides(OPT_TERRAGRUNT_OVERRIDE_ATTR))
+	}
+
+	terraformFilesInModules, err := findAllTerraformFilesInModules(terragruntOptions)
+	if err != nil {
+		return err
+	}
+
+	for _, terraformFile := range terraformFilesInModules {
+		util.Debugf(terragruntOptions.Logger, "Looking at file %s", terraformFile)
+		originalTerraformFileContents, err := util.ReadFileAsString(terraformFile)
+		if err != nil {
+			return err
+		}
+
+		updatedTerraformFileContents, codeWasUpdated, err := patchAwsProviderInTerraformCode(originalTerraformFileContents, terraformFile, terragruntOptions.AwsProviderPatchOverrides)
+		if err != nil {
+			return err
+		}
+
+		if codeWasUpdated {
+			terragruntOptions.Logger.Printf("Patching AWS provider in %s", terraformFile)
+			if err := util.WriteFileWithSamePermissions(terraformFile, terraformFile, []byte(updatedTerraformFileContents)); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// The format we expect in the .terraform/modules/modules.json file
+type TerraformModulesJson struct {
+	Modules []TerraformModule
+}
+
+type TerraformModule struct {
+	Key    string
+	Source string
+	Dir    string
+}
+
+// findAllTerraformFiles returns all Terraform source files within the modules being used by this Terragrunt
+// configuration. To be more specific, it only returns the source files downloaded for module "xxx" { ... } blocks into
+// the .terraform/modules folder; it does NOT return Terraform files for the top-level (AKA "root") module.
+//
+// NOTE: this method only supports *.tf files right now. Terraform code defined in *.json files is not currently
+// supported.d
+func findAllTerraformFilesInModules(terragruntOptions *options.TerragruntOptions) ([]string, error) {
+	// Terraform downloads modules into the .terraform/modules folder. Unfortunately, it downloads not only the module
+	// into that folder, but the entire repo it's in, which can contain lots of other unrelated code we probably don't
+	// want to touch. To find the paths to the actual modules, we read the modules.json file in that folder, which is
+	// a manifest file Terraform uses to track where the modules are within each repo. Note that this is an internal
+	// API, so the way we parse/read this modules.json file may break in future Terraform versions.
```

Comment:
Add link to https://github.com/hashicorp/terraform/blob/master/internal/modsdir/manifest.go as a reference (Note that we can't import that directly because it is labeled as internal).

Reply:
Good idea: https://github.com/gruntwork-io/terragrunt/pull/1326/commits/8fcc5060861ec81ccf754f81fb9c79248629fa7f